### PR TITLE
fix potential deadlock

### PR DESCRIPTION
**Pull request by charlesxsh**
Fix a potential deadlock at TestQueryConcurrency:

https://github.com/prometheus/prometheus/blob/664b39157381239c7c5e568ca9dcebc8732beae4/promql/engine_test.go#L59-L75

If promql/engine_test.go line 74 happened first, promql/engine_test.go line 72 will not be executed and leave sender's goroutine blocked forever after promql/engine_test.go line 75 stops .

Similar to #8964

### CODE REVIEW

**Comment by roidelapluie** [comment]
As this has never happened in practice, I'd leave this as is. It would be worse to not block if we create a bug in the tests. WDYT?

**Comment by codesome** [comment]
I am with @roidelapluie here. If it ever deadlocks, maybe it's hinting towards some issue in the actual code.

**Comment by charlesxsh** [comment]
I updated the PR with fixing potential deadlock at two places: processing and block. It can gracefully terminate the child goroutine if test exit with any issue.

**Review by codesome — COMMENTED** [review]
As stated above, we would not like to prevent these deadlocks which might hide regression or bugs in tests later on. Since this never caused any issue/flake till now, it is fine to leave it as it is.

File: promql/engine_test.go [hunk 1]
```diff
@@ -57,11 +57,21 @@ func TestQueryConcurrency(t *testing.T) {
 	defer cancelCtx()
 
 	block := make(chan struct{})
-	processing := make(chan struct{})
+	processing := make(chan struct{}, 1)
```
Comment (codesome) [inline promql/engine_test.go:60]:
```suggestion
	processing := make(chan struct{})
```

**Review by codesome — COMMENTED** [review]
Took another look and this change should be fine. But I would not increase the channel buffer size.

**Comment by charlesxsh** [comment]
@codesome updated it.

**Comment by codesome** [comment]
@charlesxsh could you please fix the DCO (it's failing now)? And possibly also sync this PR with the main branch to fix other failing tests.

**Comment by charlesxsh** [comment]
@codesome Hi there, I fixed the DCO. Would you like to merge it if everything ok?

**Review by codesome — APPROVED** [review]

**Pull request merged**